skycoin/skywire
# Improve app configurability via hypervisor and make it persistent

Fixes #66
Fixes #86
Fixes #121

Depends on https://github.com/SkycoinPro/skywire-services/pull/56

### #66

Changes:	

- If the autostart value of an application is changed by calling the `PUT /api/visors/{pk}/apps/{app} API` endpoint, the new value is restored after stopping the visor and starting it again.

How to test this PR:

1. Start some nodes using the `make integration-run-generic` command of the `skywire-services` repository.
2. Call `GET /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7`. The `autostart` property of the `skychat` will be `true`
3. Call `PUT /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7/apps/skychat` with `{autostart: false}` as content, to stop the `skychat` app.
4. Call `GET /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7` again. The `autostart` property of the `skychat` will be `false`
5. Call `make integration-teardown; tmux kill-server` in the command window that is running the `skywire-services` test enviroment, to stop it. Then call `make integration-run-generic` to start it again.
6. Call `GET /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7` again. The `autostart` property of the `skychat` will be `false`.

### #86

Changes:

- Implemented changing `skysocks` app password and `skysocks-client` app public key of a server via the `PUT /api/visors/{pk}/apps/{app} API` endpoint. 

How to test this PR:

0. Make sure https://github.com/SkycoinPro/skywire-services/pull/56 is merged. If not, you may also copy the proxy apps configs to the generic environment and try it instead of the proxy environment.
1. Start some nodes using the `make integration-run-proxy` command of the `skywire-services` repository.
2. Check `skysocks` app passcode in `integration/proxy/nodeA.json` file.
3. Call `PUT /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7/apps/skysocks` with `{"passcode": "test"}` as content, to update the passcode in the config. This should also restart the `skysocks` app.
4. Check if `skysocks` app passcode in `integration/proxy/nodeA.json` file is changed to `test`.
5. Check the `NodeA` pane in `tmux` and make sure that no error happened while restarting `skysocks`. 
6. Run `ps aux | grep skysocks` and check if `skysocks` is running.
7. Repeat `1-6` for `skysocks-client` PK (argument is named `pk`) on node `031b80cd5773143a39d940dc0710b93dcccc262a85108018a7a95ab9af734f8055`

### #121

Changes:

- Fixed the bug when visor could not run more than one app
- Fixed the bug when apps could not be restarted

How to test this PR:

1. Add the `skysocks` app config from `integration/proxy/nodeA.json` to `integration/generic/nodeA.json` as the second app.
2. Start some nodes using the `make integration-run-generic` command of the `skywire-services` repository.
3. Check `NodeA` pane in `tmux`.
4. Both apps should be started, the output of `ps aux` should show them.
5. Call `PUT /api/nodes/024ec47420176680816e0406250e7156465e4531f5b26057c9f6297bb0303558c7/apps/skysocks` with `{"passcode": "test"}` as content, to update the passcode in the config. This should also restart the `skysocks` app.
6. Check logs, there should be no error while stopping and starting `skysocks`, it also should present in the output of `ps aux | grep skysocks`.
